bpf/flows_filter.h

```diff
@@ -0,0 +1,189 @@
+/*
+    rule based filter to filter out packets not of interest to users.
+*/
+
+#ifndef __FLOWS_FILTER_H__
+#define __FLOWS_FILTER_H__
+
+#include "utils.h"
+
+// remove the comment below to enable debug prints
+//#define ENABLE_BPF_PRINTK
+#ifdef ENABLE_BPF_PRINTK
+#define BPF_PRINTK(fmt, args...) bpf_printk(fmt, ##args)
+#else
+#define BPF_PRINTK(fmt, args...)
+#endif
+
+static __always_inline int is_zero_ip(u8 *ip, u8 len) {
+    for (int i = 0; i < len; i++) {
+        if (ip[i] != 0) {
+            BPF_PRINTK("ip not zero ip[%d]:%d\n", i, ip[i]);
+            return 0;
+        }
+    }
+    return 1;
+}
+
+static __always_inline int is_equal_ip(u8 *ip1, u8 *ip2, u8 len) {
+    for (int i = 0; i < len; i++) {
+        if (ip1[i] != ip2[i]) {
+            BPF_PRINTK("ip mismatched ip1[%d]:%d not equal to ip2[%d]:%d\n", i, ip1[i], i, ip2[i]);
+            return 0;
+        }
+    }
+   return 1;
+}
+
+/*
+ * check if the flow match filter rule and return >= 1 if the flow is to be dropped
+ */
+static __always_inline int is_drop_flow(flow_id *id) {
+    struct filter_key_t key;
+    u8 len = 0, offset = 0;
+    int result = 0;
+
+    __builtin_memset(&key, 0, sizeof(key));
+    // for Ingress side we do srcIP CIDR matching and for Egress side we do dstIP CIDR matching
```

Comment:
I'm trying to see if this allows us to work with these use cases:

- I want to get all outgoing traffic from my pod `foo`
- I want to get all ingoing traffic to my pod `foo`
- I want to get all traffic (in+out) of my pod `foo`

The assumption to check srcIP for Ingress and dstIP for Egress doesn't seem to work here? E.g. let's say my pod is 1.2.3.4 and the peer is 8.8.8.8

Flow A = (src=1.2.3.4, dst=8.8.8.8, dir=egress)
Flow B = (src=8.8.8.8, dst=1.2.3.4, dir=ingress)

I'll miss these two flows if I'm filtering for CIDR (or IP) 1.2.3.4/32, no ? Because in both cases we would try to match 8.8.8.8 instead of 1.2.3.4
I know with the different interfaces involved & dups, probably I would still get something, "thanks" to the duplicates... But I'm wondering if we shouldn't instead match both SRC and DST, regardless of the direction, when doing the IP matching: that way we get all the related flows, for all involved interfaces

Reply:
huh, I just realize the rule is reversed for the IP matching. Ie for CIDR matching you check src for ingress and  dst for egress, but for IP matching you check **dst** for ingress and **src** for egress. So it's like CIDR is used to match one side of the connection and IP filter is used to match the other side?

Reply:
the idea here is as follows
- on ingress side
we match on the src CIDR if known if its not we can use `0.0.0.0/0` then u have number of options to further narrow down one of them the dstIP
- on the egress since the cluster is the one generating the traffic we filter on dst CIDR again if we don't know we can use `0.0.0.0/0` and further filter on srcIP that will become very handy when use feature like OVN EgressIP where u want to see all flows with srcIP == EgressIP

and yes this feature need doc to explain all options

Reply:
cc'd @skrthomas

Reply:
Thanks @msherif1234 just want to make sure I'm tying this in the right place, is this part of the netobserv CLI?

Reply:
> and yes this feature need doc to explain all options
I mean, the doc here in this repo must be clarified:

```go
	// FlowFilterIPCIDR is the IP CIDR to filter flows.
	// Example: 10.10.10.0/24 or 100:100:100:100::/64
	FlowFilterIPCIDR string `env:"FLOW_FILTER_IP_CIDR"`

	// FlowFilterIP is the IP to filter flows.
	// Example: 10.10.10.10
	FlowFilterIP string `env:"FLOW_FILTER_IP"`
```

=> it tells nothing about the fact that CIDR is used to filter on the remote peer (src for ingress and dst for egress), and IP filter for the local end (dst for ingress and src for egress). That's to be documented here as well.

Also there's a config.md file to update iirc

Reply:
I'm still unsure about src/dst decision that is made here regarding direction. I haven't tested but I think that's going to miss some flows. I can try testing when I find some time

Reply:
I added https://github.com/netobserv/netobserv-ebpf-agent/pull/307/commits/f2a65b3639d01c06092555312c2b85b43881aa5d to explain in details the feature and some possible use cases